### ISSUE TITLE
Validation provider doesn't decode external URIs before calling fs.readFile

### ISSUE THREAD

**Issue by lexaknyazev** [AnalyticalGraphicsInc/gltf-vscode#69]
Consider an asset like this:
```json
{
    "asset": {
        "version": "2.0"
    },
    "buffers": [
        {
            "uri":"spaces in name.bin",
            "byteLength": 1
        }
    ]
}
```

`externalResourceFunction` will be called with `spaces%20in%20name.bin` (because it's an URI), and `fs.readFile` will fail.

**Comment by donmccurdy** [comment]
Related to https://github.com/KhronosGroup/glTF/pull/1100 — Suppose instead that the asset had a "?" in the filename, and a query parameter:

```
/path/to/filename-with-%3F-question-mark.bin?and=query-params
```

It's _not_ intended that the query parameter be URI-encoded; that should be passed un-escaped to `externalResourceFunction`. The `?` in the filename _must_ be escaped in the asset, or it is invalid. So I think that technically, spaces in the URI should also be escaped in the asset itself.

Although, spaces in particular are not a big problem for the web, as this code will automatically have spaces converted to `%20` correctly:

```js
fetch('https://www.example.com/?q=here is a query')
  .then((response) => response.text());
```

**Comment by donmccurdy** [comment]
More to the point, is it an invalid URI if it contains a space?

**Comment by emackey** [comment]
Good question.  I would think ideally we'd want glTF to be somewhat tolerant, like the web is.  So an encoded space is official, but a non-encoded space is still functional.

Query parameters on relative URIs are more questionable in my mind, since that's an HTTP protocol thing, and the root URI may be `file` or `ftp` or similar with no support for query parameters.  The glTF itself may not know its own root path (hence that's passed in separately to most engines).

Obvious but worth noting that this doesn't affect "Embedded" files or the typical GLB that follows the embedded pattern, since the needed data is in the same file.  It's only an issue for files that are broken into pieces (which is sometimes an indication the file is still being edited).

**Comment by donmccurdy** [comment]
> Good question. I would think ideally we'd want glTF to be somewhat tolerant, like the web is. So an encoded space is official, but a non-encoded space is still functional.

Agreed.

**Comment by lexaknyazev** [comment]
> More to the point, is it an invalid URI if it contains a space?

It's a complex question. Officially, it shouldn't contain spaces, but there're some normalization rules to fix that on the fly.

This issue is aimed at fixing `fs.readFile` call.

**Comment by donmccurdy** [comment]
Sure. I have no problem with the PR, feel free to merge that. But (now or later) I think we should think through whether the validator should be URI-encoding what is already supposed to be a valid URI.

**Comment by emackey** [comment]
Fixed in `2.1.2`
